### PR TITLE
Adding autocodeblocking

### PR DESCRIPTION
**Pull request by hargoniX**
Autodetecting python code via checking if
* it has at least three lines
* is not throwing syntax errors when parsed with ast.parse
* is not only ast.Expr

### CODE REVIEW

File: bot/cogs/bot.py [hunk 1]
```diff
@@ -60,6 +62,18 @@ def __init__(self, bot: AutoShardedBot):
 
         await ctx.invoke(self.info)
 
+    async def on_message(self, msg: Message):
+        """
+        Auto codeblock python code
+        """
+        if msg.content.count("\n") >= 3:  # more than three lines
```
Comment (JoeBanks13) [inline bot/cogs/bot.py]: Didn't we agree that it was over 3?

Comment (AnonGuy) [reply bot/cogs/bot.py]: Three or more counts of `\n` should mean over 3 lines, unless the user includes trailing newlines.

**Comment by hargoniX** [comment]
Right gimme few secs drinking

File: bot/cogs/bot.py (hunk 1 above)
Comment (JoeBanks13) [reply bot/cogs/bot.py]: Ahh yeah, my bad

**Comment by hargoniX** [comment]
Oh uuhh it ok the I guess (didn't see aperture

File: bot/cogs/bot.py [hunk 2]
```diff
@@ -60,6 +62,18 @@ def __init__(self, bot: AutoShardedBot):
 
         await ctx.invoke(self.info)
 
+    async def on_message(self, msg: Message):
+        """
+        Auto codeblock python code
```
Comment (JoeBanks13) [inline bot/cogs/bot.py]: Is this docstring necessary? In future it may not just be used for parsing Python code.

File: bot/cogs/bot.py [hunk 3]
```diff
@@ -60,6 +62,18 @@ def __init__(self, bot: AutoShardedBot):
...
+        """
+        if msg.content.count("\n") >= 3:  # more than three lines
+            try:
+                tree = ast.parse(msg.content)  # no syntax errors
+                if not all(isinstance(node, ast.Expr) for node in tree.body):  # we dont want hi\nthere\nguys\nD
+                    await self.bot.send_message(msg.author, "```Python\n{}\n```".format(msg.content))
```
Comment (JoeBanks13) [inline bot/cogs/bot.py]: Should technically be a lower case `p`, as well as this we should probably either delete the message, or, have it send the current text in the codeblock tag.

File: bot/cogs/bot.py (hunk 2 above)
Comment (hargoniX) [reply bot/cogs/bot.py]: Yes, one could, of course, delete it but just for current usage. Depends on you del, not del?

File: bot/cogs/bot.py (hunk 3 above)
Comment (hargoniX) [reply bot/cogs/bot.py]: ok

Comment (JoeBanks13) [reply bot/cogs/bot.py]: Regarding previous comment, the lowercase `p` was referring to the syntax block.
```py
"```python\n..."
```

File: bot/cogs/bot.py [hunk 4]
```diff
@@ -60,6 +62,18 @@ def __init__(self, bot: AutoShardedBot):
...
         await ctx.invoke(self.info)
 
+    async def on_message(self, msg: Message):
+        """
+        Auto codeblock python code
+        """
+        if msg.content.count("\n") >= 3:  # more than three lines
+            try:
+                tree = ast.parse(msg.content)  # no syntax errors
+                if not all(isinstance(node, ast.Expr) for node in tree.body):  # we dont want hi\nthere\nguys\nD
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: This inline comment seems unnecessary, maybe have a comment underneath describing the check more clearly?

Comment (JoeBanks13) [reply bot/cogs/bot.py]: Inline comments never really look nice, and a couple here are probably un-needed, especially after the if statement and the tree definition. Moving them above or underneath a line, and adding newlines around that is a better idea.

Comment (hargoniX) [reply bot/cogs/bot.py]: alright

File: bot/cogs/bot.py (hunk 3 above)
Comment (1mn) [reply bot/cogs/bot.py]: We use f-strings all over the bot, this should also be an f-string.

File: bot/cogs/bot.py [hunk 5]
```diff
@@ -60,6 +62,20 @@ def __init__(self, bot: AutoShardedBot):
 
         await ctx.invoke(self.info)
 
+    async def on_message(self, msg: Message):
```
Comment (shadycake) [inline bot/cogs/bot.py]: Is having this directly in bot.py a good idea? It might be better to have it elsewhere and be imported, for readability. ( And have that docstring moved there )
But thats open to discussion.

File: bot/cogs/bot.py (hunk 2 above)
Comment (JoeBanks13) [reply bot/cogs/bot.py]: If you look at the other files where event listeners are defined they don't have docstrings, to stay compliant, this should be removed.

File: bot/cogs/bot.py (hunk 4 above)
Comment (1mn) [reply bot/cogs/bot.py]: Your comments should follow PEP8 - they should start with a capital letter, and need to be grammatically correct. That means you're gonna need to use some punctuation. This comment, for example, deseperately needs a period and a comma.

File: bot/cogs/bot.py [hunk 6]
```diff
@@ -60,6 +62,17 @@ def __init__(self, bot: AutoShardedBot):
...
+        if msg.content.count("\n") >= 3:  # more than three lines
+            try:
+                tree = ast.parse(msg.content)
+                # tries to parse the code into a valid ast if it throws a SyntaxError it is no valid python
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: May I suggest `# Attempts to parse the message into an AST node.`
`# Invalid Python code will raise a SyntaxError.`

File: bot/cogs/bot.py [hunk 7]
```diff
@@ -60,6 +62,18 @@ def __init__(self, bot: AutoShardedBot):
 
         await ctx.invoke(self.info)
 
+    async def on_message(self, msg: Message):
+        if msg.content.count("\n") >= 3:  # more than three lines
+            try:
+                tree = ast.parse(msg.content)
+                # Attempts to parse the message into an AST node.
+                # Invalid Python code will raise a SyntaxError.
+                if not all(isinstance(node, ast.Expr) for node in tree.body):  # we dont want hi\nthere\nguys\nD
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: This inline comment still bugs me. Try moving it to a new line and changing the comment to be more descriptive?

File: bot/cogs/bot.py [hunk 8]
```diff
@@ -60,6 +62,18 @@ def __init__(self, bot: AutoShardedBot):
...
+                    await self.bot.send_message(msg.author, "```python\n{}\n```".format(msg.content))
```
Comment (shadycake) [inline bot/cogs/bot.py]: send_message() is pre-rewrite, and I'm pretty sure you want to send it to the channel, so the rewrite version would be `msg.channel.send(<message here>)`
Also if the message (with the unformatted python code) were >1986 characters this line would throw an error, maybe add a check against this?

File: bot/cogs/bot.py [hunk 9]
```diff
@@ -60,6 +62,20 @@ def __init__(self, bot: AutoShardedBot):
...
 
+    async def on_message(self, msg: Message):
+        if msg.content.count("\n") >= 3:  # more than three lines
+            try:
+                tree = ast.parse(msg.content)
+
+                # Attempts to parse the message into an AST node.
+                # Invalid Python code will raise a SyntaxError.
+                # => text which is not Python code will be filtered.
+                if not all(isinstance(node, ast.Expr) for node in tree.body):  # we dont want hi\nthere\nguys\nD
```
Comment (1mn) [inline bot/cogs/bot.py]: is this inline comment still necessary?

Comment (AnonGuy) [inline bot/cogs/bot.py]: `# we dont want hi\nthere\nguys\nD` is what I meant. Try removing it and adding a new comment underneath

File: bot/cogs/bot.py [hunk 10]
```diff
@@ -60,6 +62,20 @@ def __init__(self, bot: AutoShardedBot):
...
+                    await self.bot.send_message(msg.author, "```python\n{}\n```".format(msg.content))
```
Comment (1mn) [inline bot/cogs/bot.py]: this should be an f-string to be consistent with the rest of the repo.

File: bot/cogs/bot.py [hunk 11]
```diff
@@ -60,6 +62,20 @@ def __init__(self, bot: AutoShardedBot):
 
         await ctx.invoke(self.info)
 
+    async def on_message(self, msg: Message):
+        if msg.content.count("\n") >= 3:  # more than three lines
```
Comment (1mn) [inline bot/cogs/bot.py]: this inline comment states the obvious, which PEP8 specifically says not to do.

File: bot/cogs/bot.py (hunk 9 above)
Comment (AnonGuy) [reply bot/cogs/bot.py]: Something like:
`# Single-line words are interpreted as expressions.`
`# This checks if all nodes were parsed as expressions.`

File: bot/cogs/bot.py [hunk 12]
```diff
@@ -60,6 +62,20 @@ def __init__(self, bot: AutoShardedBot):
...
+    async def on_message(self, msg: Message):
+        """
+        Auto codeblock python code
+        """
+        if msg.content.count("\n") >= 3:  # more than three lines
+            try:
+                tree = ast.parse(msg.content)
+                # tries to parse the code into a valid ast if it throws a SyntaxError it is no valid python
+                if not all(isinstance(node, ast.Expr) for node in tree.body):  # we dont want hi\nthere\nguys\nD
+                    await self.bot.send_message(msg.author, "```python\n{}\n```".format(msg.content))
```
Comment (shadycake) [inline bot/cogs/bot.py]: send_message() is pre-rewrite, and I'm pretty sure you want to send it to the channel, so the rewrite version would be msg.channel.send(<message here>)
Also if the message (with the unformatted python code) were >1986 characters this line would throw an error, maybe add a check against this?
(reposting this comment because the old one shows up as outdated)

File: bot/cogs/bot.py (hunk 10 above)
Comment (shadycake) [inline bot/cogs/bot.py]: send_message() is pre-rewrite, and I'm pretty sure you want to send it to the channel, so the rewrite version would be msg.channel.send(<message here>)
Also if the message (with the unformatted python code) were >1986 characters this line would throw an error, maybe add a check against this?
(adding again because previous comment is marked as out of date, when the line hasn't been fixed)

File: bot/cogs/bot.py [hunk 13]
```diff
@@ -60,6 +62,27 @@ def __init__(self, bot: AutoShardedBot):
...
 
+    async def on_message(self, msg: Message):
+        if msg.content.count("\n") >= 3:  # more than three lines
+            try:
+                tree = ast.parse(msg.content)
+                # Attempts to parse the message into an AST node.
+                # Invalid Python code will raise a SyntaxError.
+                if not all(isinstance(node, ast.Expr) for node in tree.body):
+                    # we dont want multiple lines of single words like:
+                    '''
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: Is a docstring really necessary here?

Comment (hargoniX) [reply bot/cogs/bot.py]: How else would one do multiline comments without 
# 
# 
# 
or should i use that?

File: bot/cogs/bot.py [hunk 14]
```diff
@@ -60,6 +62,27 @@ def __init__(self, bot: AutoShardedBot):
...
+                    hi
+                    there
+                    guys
+                    '''
+                    # that would be syntactically valid Python but in this case
+                    # just some random multiline text someone is sending.
+
+                    await msg.channel.send("```python\n{}\n```".format(msg.content))
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: This line still needs an f-string...

File: bot/cogs/bot.py (hunk 13 above)
Comment (AnonGuy) [reply bot/cogs/bot.py]: You don't really need an example. I think just explaining what the check does should suffice.

File: bot/cogs/bot.py (hunk 14 above)
Comment (shadycake) [inline bot/cogs/bot.py]: If msg.content were >1986 characters long, discord will get angry at you as the message exceeds 2000 characters, to avoid unnecessary errors I suggest some sort of check against this? Maybe format the string, then check if it is over 2000 characters before sending and deleting

File: bot/cogs/bot.py (hunk 13 above)
Comment (1mn) [reply bot/cogs/bot.py]: yeah don't use docstrings like this. if you did need a comment of that many lines, you should have used multiple # comments, but I agree with Ap that you don't really need that in this case. Just say it with words.

File: bot/cogs/bot.py [hunk 15]
```diff
@@ -72,9 +72,10 @@ def __init__(self, bot: AutoShardedBot):
                     # we dont want multiple lines of single words,
                     # they would be syntactically valid Python but could also be
                     # just some random multiline text someone is sending.
-
-                    await msg.channel.send(f"```python\n{msg.content}\n```")
-                    await self.bot.delete_message(msg)
+                    formatted = f"```python\n{msg.content}\n```"
+                    if len(formatted) > 2000:
```
Comment (shadycake) [inline bot/cogs/bot.py]: Less than or equal to 2000, not more

File: bot/cogs/bot.py [hunk 16]
```diff
@@ -60,6 +62,23 @@ def __init__(self, bot: AutoShardedBot):
...
+                # Attempts to parse the message into an AST node.
+                # Invalid Python code will raise a SyntaxError.
+                if not all(isinstance(node, ast.Expr) for node in tree.body):
+                    # we dont want multiple lines of single words,
+                    # they would be syntactically valid Python but could also be
+                    # just some random multiline text someone is sending.
+                    formatted = f"```python\n{msg.content}\n```"
+                    if len(formatted) > 2000:
+                        await msg.channel.send(formatted)
+                        await self.bot.delete_message(msg)
```
Comment (JoeBanks13) [inline bot/cogs/bot.py]: This is still async discord.py

File: bot/cogs/bot.py [hunk 17]
```diff
@@ -60,6 +62,23 @@ def __init__(self, bot: AutoShardedBot):
...
+            try:
+                tree = ast.parse(msg.content)
+                # Attempts to parse the message into an AST node.
+                # Invalid Python code will raise a SyntaxError.
+                if not all(isinstance(node, ast.Expr) for node in tree.body):
+                    # we dont want multiple lines of single words,
+                    # they would be syntactically valid Python but could also be
+                    # just some random multiline text someone is sending.
+                    formatted = f"```python\n{msg.content}\n```"
+                    if len(formatted) > 2000:
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: I think you mean `if len(formatted) <= 2000`.

Comment (JoeBanks13) [reply bot/cogs/bot.py]: Yeah, this will make the bot only send if over 2000 chars...

File: bot/cogs/bot.py (hunk 16 above)
Comment (hargoniX) [reply bot/cogs/bot.py]: What do you mean?

File: bot/cogs/bot.py (hunk 5 above)
Comment (1mn) [reply bot/cogs/bot.py]: I think it's fine for now.

File: bot/cogs/bot.py (hunk 16 above)
Comment (shadycake) [reply bot/cogs/bot.py]: for reference: it would be `await bot.delete()`

Comment (JoeBanks13) [reply bot/cogs/bot.py]: No, `await msg.delete()`, addressed it in my requested changes.

Comment (hargoniX) [reply bot/cogs/bot.py]: uuh @shadycake bot.delete() made snekchek to show an error

File: bot/cogs/bot.py [hunk 18]
```diff
@@ -75,7 +75,7 @@ def __init__(self, bot: AutoShardedBot):
                     formatted = f"```python\n{msg.content}\n```"
                     if len(formatted) <= 2000:
                         await msg.channel.send(formatted)
-                        await self.bot.delete_message(msg)
+                        await bot.delete_message(msg)
```
Comment (JoeBanks13) [inline bot/cogs/bot.py]: It's `await msg.delete()`, in discord.py rewrite, all the actions upon an item are relative to the item itself.

File: bot/cogs/bot.py [hunk 19]
```diff
@@ -60,6 +62,23 @@ def __init__(self, bot: AutoShardedBot):
...
+    async def on_message(self, msg: Message):
+        if msg.content.count("\n") >= 3:
+            try:
+                tree = ast.parse(msg.content)
+                # Attempts to parse the message into an AST node.
+                # Invalid Python code will raise a SyntaxError.
+                if not all(isinstance(node, ast.Expr) for node in tree.body):
+                    # we dont want multiple lines of single words,
+                    # they would be syntactically valid Python but could also be
+                    # just some random multiline text someone is sending.
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: This is pretty minor, but in future separate and capitalise sentences when commenting.
Refer to the comments above this one.

Comment (1mn) [reply bot/cogs/bot.py]: the comment content itself is pretty good now though.

**Comment by 1mn** [comment]
Okay, this PR is starting to come together. Can we have a blank line before each comment block? it's a bit dense right now.

File: bot/cogs/bot.py [hunk 20]
```diff
@@ -60,6 +62,23 @@ def __init__(self, bot: AutoShardedBot):
...
+                    formatted = f"```python\n{msg.content}\n```"
+                    if len(formatted) <= 2000:
+                        await msg.channel.send(formatted)
+                        await msg.delete()
+            except SyntaxError:
+                pass
+
```
Comment (shadycake) [inline bot/cogs/bot.py]: `await self.bot.process_commands()`
We don't want our beautiful commands to not work now do we?

File: bot/cogs/bot.py [hunk 21]
```diff
@@ -60,6 +62,23 @@ def __init__(self, bot: AutoShardedBot):
...
         await ctx.invoke(self.info)
 
+    async def on_message(self, msg: Message):
+        if msg.content.count("\n") >= 3:
+            try:
+                tree = ast.parse(msg.content)
+                # Attempts to parse the message into an AST node.
+                # Invalid Python code will raise a SyntaxError.
+                if not all(isinstance(node, ast.Expr) for node in tree.body):
+                    # We dont want multiple lines of single words,
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: don't*

File: bot/cogs/bot.py [hunk 22]
```diff
@@ -60,6 +62,23 @@ def __init__(self, bot: AutoShardedBot):
...
+            try:
+                tree = ast.parse(msg.content)
+                # Attempts to parse the message into an AST node.
+                # Invalid Python code will raise a SyntaxError.
+                if not all(isinstance(node, ast.Expr) for node in tree.body):
+                    # we dont want multiple lines of single words,
+                    # they would be syntactically valid Python but could also be
+                    # just some random multiline text someone is sending.
+                    formatted = f"```python\n{msg.content}\n```"
+                    if len(formatted) <= 2000:
```
Comment (JoeBanks13) [inline bot/cogs/bot.py]: Not going to start this as a requested change but do we want the bot to replace the users content, or do we want the bot to send a message informing the user of how to code block?

File: bot/cogs/bot.py [hunk 23]
```diff
@@ -60,6 +62,25 @@ def __init__(self, bot: AutoShardedBot):
...
+    async def on_message(self, msg: Message):
+        if msg.content.count("\n") >= 3:
+            try:
+                tree = ast.parse(msg.content)
+
+                # Attempts to parse the message into an AST node.
+                # Invalid Python code will raise a SyntaxError.
+                if not all(isinstance(node, ast.Expr) for node in tree.body):
+
+                    # We don't* want multiple lines of single words,
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: You don't need this `*`

File: bot/cogs/bot.py (hunk 22 above)
Comment (hargoniX) [reply bot/cogs/bot.py]: i believe you cannot edit other peoples msgs on discord can you?

File: bot/cogs/bot.py (hunk 20 above)
Comment (shadycake) [reply bot/cogs/bot.py]: Actually put that at the start of on_message, not afterwards

Comment (JoeBanks13) [reply bot/cogs/bot.py]: I was in belief that cog event listeners don't stop the bot from processing commands? Otherwise all on_message events would have to call that. Correct me if I'm wrong.

File: bot/cogs/bot.py (hunk 22 above)
Comment (hargoniX) [reply bot/cogs/bot.py]: I would say replace because some user will take some time to actually understand how to to use codeblock

Comment (1mn) [reply bot/cogs/bot.py]: we want to inform the user. The goal is for the user to learn how to do this in the future.

File: bot/cogs/bot.py (hunk 20 above)
Comment (shadycake) [reply bot/cogs/bot.py]: on_message usually calls process_commands(), but he's overwriting on_message meaning he has to manually put process_commands there, I had to do the same with my bot (When doing everyone's favourite XP system).

File: bot/cogs/bot.py (hunk 22 above)
Comment (JoeBanks13) [reply bot/cogs/bot.py]: They seem to grasp our current `py!tag codeblock` text soon enough, and I think it's better letting them know for future.

Comment (1mn) [reply bot/cogs/bot.py]: I actually think this is a critical point. replacing their message teaches them nothing. they'll go "oh cool" and have zero motivation to learn how to do it.

File: bot/cogs/bot.py (hunk 20 above)
Comment (hargoniX) [reply bot/cogs/bot.py]: But bot.process_commands() will cause errors again wont it?

Comment (JoeBanks13) [reply bot/cogs/bot.py]: @gdude2002 Hey, sorry to pull you into this, happen to know whether a cogs `on_message` function needs a `self.bot.process_commands(msg)` call?

Comment (gdude2002) [reply bot/cogs/bot.py]: Er, it doesn't. We have several that don't have it.

Comment (JoeBanks13) [reply bot/cogs/bot.py]: Just clarified with gdude, it's not required, ignore the process_commands function, un-needed.

Comment (shadycake) [reply bot/cogs/bot.py]: Ok my bad I was wrong

Comment (hargoniX) [reply bot/cogs/bot.py]: Yes, i can read :wink:.

**Comment by hargoniX** [comment]
Lemon?

File: bot/cogs/bot.py (hunk 22 above)
Comment (JoeBanks13) [reply bot/cogs/bot.py]: Text from current tag:

If you are going to post code in Discord, please use syntax highlighted blocks, as it makes it more legible for other users.

To do this, you should input your content like this:

\\\`\\\`\\\`python
print("Hello world!")
\\\`\\\`\\\`

This will result in the following:
\`\`\`
print("Hello world!")
\`\`\`

**Comment by hargoniX** [comment]
Like pybutt with an embed?

**Comment by JoeBanks13** [comment]
Preferably with, title being something like "Code formatting" and content is the content I've linked in my requested changes.

**Comment by JoeBanks13** [comment]
GitHub seems to have hidden my changes somewhere, the text from the current tag is here:

If you are going to post code in Discord, please use syntax highlighted blocks, as it makes it more legible for other users.

To do this, you should input your content like this:

\\\`\\\`\\\`python
print("Hello world!")
\\\`\\\`\\\`

This will result in the following:
\`\`\`
print("Hello world!")
\`\`\`

**Comment by 1mn** [comment]
If we're gonna post this embed, it also needs a cooldown so it doesn't spam. Maybe 5 minute cooldown per channel would be okay.

**Comment by JoeBanks13** [comment]
As well as this, I was thinking of adding a restriction to only enforce code blocking inside help channels, and *maybe* the python channel.

**Comment by hargoniX** [comment]
Shall i do this via channel name or ID?

**Comment by JoeBanks13** [comment]
ID, that way if anything changes in future we're still good.

**Comment by hargoniX** [comment]
@JoeBanks13 how do i store the embed line better?

File: bot/cogs/bot.py [hunk 24]
```diff
@@ -60,6 +65,40 @@ def __init__(self, bot: AutoShardedBot):
...
+print("Hello world!")
+\`\`\`
+
+This will result in the following:
+```
+print("Hello world!")
+```
+"""
+                            information = Embed(title="Code formatting")
+                            information.add_field(name="Information", value=howto)
```
Comment (JoeBanks13) [inline bot/cogs/bot.py]: Why not just `Embed(title="Code formatting", description=howto)`

Comment (hargoniX) [reply bot/cogs/bot.py]: Because i did not know about that

File: bot/cogs/bot.py [hunk 25]
```diff
@@ -15,6 +18,8 @@ class Bot:
 
     def __init__(self, bot: AutoShardedBot):
         self.bot = bot
+        self.allowed = {303906576991780866: 0, 303906556754395136: 0, 303906514266226689: 0, 267624335836053506: 0}
```
Comment (1mn) [inline bot/cogs/bot.py]: this is not a very descriptive variable name for this attribute.

**Comment by hargoniX** [comment]
I am missing one?

File: bot/cogs/bot.py [hunk 26]
```diff
@@ -15,6 +18,8 @@ class Bot:
...
+        # stores allowed channels plus unix timestmp from last call
```
Comment (1mn) [inline bot/cogs/bot.py]: comments do not go after the line they apply to. missing capital S. timestmp isn't a word.

File: bot/cogs/bot.py [hunk 27]
```diff
@@ -60,6 +65,39 @@ def __init__(self, bot: AutoShardedBot):
...
+
+\`\`\`python
+print("Hello world!")
+\`\`\`
+
+This will result in the following:
+```
+print("Hello world!")
+```
+"""
```
Comment (1mn) [inline bot/cogs/bot.py]: I hate this method, because of the indentation. It just looks messy to me. Use an implicitely wrapped string: 

```py
    howto = ("Please use syntax highlighted blocks, as it makes it more legible for other users.\n"
             "\n"
             "To do this, you should input your content like this:\n"
             "... and so on"
    )
```

If someone violently disagrees with this, okay fine I'll budge. but that's my personal preference for this somewhat tricky readability problem.

File: bot/cogs/bot.py (hunk 25 above)
Comment (1mn) [reply bot/cogs/bot.py]: I'd prefer multilining that dict as well.

```py
self.better_variable_name = {
    303906576991780866: 0, 
    303906556754395136: 0
}
```

File: bot/cogs/bot.py (hunk 27 above)
Comment (hargoniX) [reply bot/cogs/bot.py]: well it causes some line length errors

Comment (1mn) [reply bot/cogs/bot.py]: yep, you still can't exceed 120. If you're in danger of doing that, simply wrap that line onto the next line and don't put in the newline.

File: bot/cogs/bot.py [hunk 28]
```diff
@@ -60,6 +65,39 @@ def __init__(self, bot: AutoShardedBot):
...
+                        tree = ast.parse(msg.content)
+
+                        # Attempts to parse the message into an AST node.
+                        # Invalid Python code will raise a SyntaxError.
+                        if not all(isinstance(node, ast.Expr) for node in tree.body):
+
+                            # We don't want multiple lines of single words,
+                            # They would be syntactically valid Python but could also be
+                            # Just some random multiline text someone is sending.
+                            howto = ("Please use syntax highlighted blocks, as it makes it more legible for other users.\n"
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: I would prefer sticking these floating `"\n"`s back on the existing lines.

File: bot/cogs/bot.py [hunk 29]
```diff
@@ -60,6 +65,36 @@ def __init__(self, bot: AutoShardedBot):
...
+                        # Attempts to parse the message into an AST node.
+                        # Invalid Python code will raise a SyntaxError.
+                        if not all(isinstance(node, ast.Expr) for node in tree.body):
+
+                            # We don't want multiple lines of single words,
+                            # They would be syntactically valid Python but could also be
+                            # Just some random multiline text someone is sending.
+                            howto = ("Please use syntax highlighted blocks, as it makes it more legible for other users.\n\n"
+                                     "To do this, you should input your content like this:\n\n"
+                                     "\`\`\`python\n"
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: equal `\n`s pl0x

File: bot/cogs/bot.py [hunk 30]
```diff
@@ -60,6 +65,36 @@ def __init__(self, bot: AutoShardedBot):
 
         await ctx.invoke(self.info)
 
+    async def on_message(self, msg: Message):
+        if msg.channel.id in self.code_block_channels:
+            if self.allowed[msg.channel.id]-time.time() > 300:
```
Comment (1mn) [inline bot/cogs/bot.py]: NameError? come on dude, at least lint out the syntax errors before you commit.

Comment (hargoniX) [reply bot/cogs/bot.py]: http://prntscr.com/ijpqmb my plugin didnt see it :(

File: bot/cogs/bot.py [hunk 31]
```diff
@@ -15,6 +18,8 @@ class Bot:
 
     def __init__(self, bot: AutoShardedBot):
         self.bot = bot
+        self.code_block_channels = {303906576991780866: 0, 303906556754395136: 0, 303906514266226689: 0, 267624335836053506: 0}
+        # Stores allowed channels plus unix timestamp from last call
```
Comment (1mn) [inline bot/cogs/bot.py]: This comment is still in the wrong place. Comments should not be placed after the line they document.

File: bot/cogs/bot.py [hunk 32]
```diff
@@ -15,6 +18,8 @@ class Bot:
 
     def __init__(self, bot: AutoShardedBot):
         self.bot = bot
+        self.code_block_channels = {303906576991780866: 0, 303906556754395136: 0, 303906514266226689: 0, 267624335836053506: 0}
```
Comment (1mn) [inline bot/cogs/bot.py]: I still want this line multilined.

File: bot/cogs/bot.py [hunk 33]
```diff
@@ -60,6 +65,36 @@ def __init__(self, bot: AutoShardedBot):
...
+            if self.code_block_channels[msg.channel.id]-time.time() > 300:
+                if msg.content.count("\n") >= 3:
+                    try:
+                        tree = ast.parse(msg.content)
+
+                        # Attempts to parse the message into an AST node.
+                        # Invalid Python code will raise a SyntaxError.
+                        if not all(isinstance(node, ast.Expr) for node in tree.body):
+
+                            # We don't want multiple lines of single words,
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: Can I suggest having:
`# Multiple lines of single words could be interpreted as expressions.`
`# This check is to avoid all nodes being parsed as expressions.`
`# (e.g. words over multiple lines)`

File: bot/cogs/bot.py [hunk 34]
```diff
@@ -60,6 +65,36 @@ def __init__(self, bot: AutoShardedBot):
...
+                        tree = ast.parse(msg.content)
+
+                        # Attempts to parse the message into an AST node.
+                        # Invalid Python code will raise a SyntaxError.
+                        if not all(isinstance(node, ast.Expr) for node in tree.body):
+
+                            # Multiple lines of single words could be interpreted as expressions.
+                            # This check is to avoid all nodes being parsed as expressions.
+                            # (e.g. words over multiple lines)
+                            howto = ("Please use syntax highlighted blocks, as it makes it more legible for other users.\n"
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: This string should be split over two lines. It exceeds 120 characters. Also made a minor edit to the sentence :D
```
"Please use syntax highlighted blocks, as it makes"
"your code more legible for other users.\n"
```

File: bot/cogs/bot.py (hunk 32 above)
Comment (AnonGuy) [reply bot/cogs/bot.py]: I agree. This line is way too long.

File: bot/cogs/bot.py [hunk 35]
```diff
@@ -16,6 +19,13 @@ class Bot:
     def __init__(self, bot: AutoShardedBot):
         self.bot = bot
 
+        # Stores allowed channels plus unix timestamp from last call
+        self.code_block_channels = {303906576991780866: 0,
+                                    303906556754395136: 0,
+                                    303906514266226689: 0,
+                                    267624335836053506: 0
+        }
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: Try moving this closing parenthesis to right after the `0`.
Reason: Travis

File: bot/cogs/bot.py [hunk 36]
```diff
@@ -1,10 +1,13 @@
...
+import ast
+import time
+
+from discord import Embed, Message
 from discord.ext.commands import AutoShardedBot, Context, command, group
 
 from dulwich.repo import Repo
 
-from bot.constants import PYTHON_GUILD, VERIFIED_ROLE
+from bot.constants import HELP1_CHANNEL, HELP2_CHANNEL, HELP3_CHANNEL, PYTHON_CHANNEL, PYTHON_GUILD, VERIFIED_ROLE
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: You know what to do.

Comment (AnonGuy) [reply bot/cogs/bot.py]: Either have two imports on each line, or have three on each line. It'll make things look nicer, promise!

Comment (hargoniX) [reply bot/cogs/bot.py]: hmmmmmmmmmmmmmmmmmmmmmmmmm

File: bot/constants.py [hunk 37]
```diff
@@ -7,6 +7,10 @@
 BOT_CHANNEL = 267659945086812160
 DEVLOG_CHANNEL = 409308876241108992
 VERIFICATION_CHANNEL = 352442727016693763
+HELP1_CHANNEL = 303906576991780866
+HELP2_CHANNEL = 303906556754395136
+HELP3_CHANNEL = 303906514266226689
+PYTHON_CHANNEL = 267624335836053506
```
Comment (AnonGuy) [inline bot/constants.py]: Can we change this order to:
```
BOT_CHANNEL
HELP1_CHANNEL
HELP2_CHANNEL
HELP3_CHANNEL
PYTHON_CHANNEL
DEVLOG_CHANNEL
VERIFICATION_CHANNEL
```
To keep line lengths ascending, thanks

File: bot/cogs/bot.py (hunk 36 above)
Comment (AnonGuy) [reply bot/cogs/bot.py]: Can we have an equal number of imports spread out over either `2` or `3` lines.
Currently, the line looks too long for use in a production setting.

File: bot/cogs/bot.py [hunk 38]
```diff
@@ -1,10 +1,14 @@
...
+import time
+
+from discord import Embed, Message
 from discord.ext.commands import AutoShardedBot, Context, command, group
 
 from dulwich.repo import Repo
 
-from bot.constants import PYTHON_GUILD, VERIFIED_ROLE
+from bot.constants import HELP1_CHANNEL, HELP2_CHANNEL, HELP3_CHANNEL
+from bot.constants import PYTHON_CHANNEL, PYTHON_GUILD, VERIFIED_ROLE
```
Comment (AnonGuy) [inline bot/cogs/bot.py]: There's no need to have two `import` statements. You can spread the imports over `2` or `3` lines without repeating `import` statements, by using surrounding parentheses.

File: bot/constants.py [hunk 39]
```diff
@@ -5,9 +5,14 @@
...
 
 BOT_CHANNEL = 267659945086812160
+HELP1_CHANNEL = 303906576991780866
+HELP2_CHANNEL = 303906556754395136
+HELP3_CHANNEL = 303906514266226689
+PYTHON_CHANNEL = 267624335836053506
 DEVLOG_CHANNEL = 409308876241108992
 VERIFICATION_CHANNEL = 352442727016693763
 
+
```
Comment (1mn) [inline bot/constants.py]: why is this here

File: bot/cogs/bot.py (hunk 38 above)
Comment (1mn) [reply bot/cogs/bot.py]: this I don't agree with. multiple imports are fine.

Comment (AnonGuy) [reply bot/cogs/bot.py]: understandable, have a nice day

File: bot/constants.py (hunk 39 above)
Comment (AnonGuy) [reply bot/constants.py]: This line is unacceptable. Requested changes.

**Comment by 1mn** [comment]
We've murdered this PR and you put up with it. Well done. The code looks great now.

**Comment by hargoniX** [comment]
hurray
